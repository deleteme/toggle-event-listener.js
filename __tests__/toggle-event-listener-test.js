import toggleEventListener from '../toggle-event-listener.js';

describe('toggleEventListener()', () => {
  it('should return the element', () => {
    expect.assertions(1);
    const element = document.createElement('div');
    const handler = jest.fn();
    expect(toggleEventListener(element, 'click', handler)).toBe(element);
  });

  it('should add the event handler if one is not already added.', () => {
    expect.assertions(1);
    const element = document.createElement('div');
    const handler = jest.fn();
    toggleEventListener(element, 'click', handler);
    element.click();
    expect(handler).toHaveBeenCalledTimes(1);
  });
  it('should be able to add handlers of different types.', () => {
    expect.assertions(2);
    const element = document.createElement('div');
    const handleClick = jest.fn();
    const handleMouseMove = jest.fn();
    toggleEventListener(element, 'click', handleClick);
    toggleEventListener(element, 'mousemove', handleMouseMove);
    element.click();
    expect(handleClick).toHaveBeenCalledTimes(1);
    expect(handleMouseMove).toHaveBeenCalledTimes(0);
  });
  it('should remove the event handler if one is already added.', () => {
    expect.assertions(1);
    const element = document.createElement('div');
    const handler = jest.fn();
    toggleEventListener(element, 'click', handler);
    toggleEventListener(element, 'click', handler);
    element.click();
    expect(handler).toHaveBeenCalledTimes(0);
  });

  describe('with declarative truthy state', () => {
    it('should add an event handler', () => {
      expect.assertions(1);
      const element = document.createElement('div');
      const handler = jest.fn();
      toggleEventListener(element, 'click', handler, true);
      element.click();
      expect(handler).toHaveBeenCalledTimes(1);
    });
    it('should only add one event handler', () => {
      expect.assertions(1);
      const element = document.createElement('div');
      const handler = jest.fn();
      toggleEventListener(element, 'click', handler, true);
      toggleEventListener(element, 'click', handler, true);
      element.click();
      expect(handler).toHaveBeenCalledTimes(1);
    });
  });
  describe('with declarative falsey state', () => {
    it('should remove an event handler', () => {
      expect.assertions(1);
      const element = document.createElement('div');
      const handler = jest.fn();
      toggleEventListener(element, 'click', handler, true);
      toggleEventListener(element, 'click', handler, false);
      toggleEventListener(element, 'click', handler, 0);
      element.click();
      expect(handler).toHaveBeenCalledTimes(0);
    });
    it('should not add the handler', () => {
      expect.assertions(1);
      const element = document.createElement('div');
      const handler = jest.fn();
      toggleEventListener(element, 'click', handler, false);
      element.click();
      expect(handler).toHaveBeenCalledTimes(0);
    });
  });
});
