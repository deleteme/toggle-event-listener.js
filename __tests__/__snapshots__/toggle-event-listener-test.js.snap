// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toggleEventListener() without required arguments should throw an error. 1`] = `"Expected event type to be a string"`;

exports[`toggleEventListener() without required arguments should throw an error. 2`] = `"Expected event type to be a string"`;

exports[`toggleEventListener() without required arguments should throw an error. 3`] = `"Expected event handler to be a function"`;

exports[`toggleEventListener() without required arguments should throw an error. 4`] = `"Invalid value used as weak map key"`;

exports[`toggleEventListener() without required arguments should throw an error. 5`] = `"Expected event type to be a string"`;

exports[`toggleEventListener() without required arguments should throw an error. 6`] = `"Expected event handler to be a function"`;
